CIScripts/Build/BuildFunctions.ps1

```diff
@@ -401,3 +401,58 @@ function Invoke-AgentTestsBuild {
 
     $Job.PopStep()
 }
+
+function Invoke-ContainerBuild {
+    Param ([Parameter(Mandatory = $true)] [string] $WorkDir,
+           [Parameter(Mandatory = $true)] [string] $ContainerSuffix,
+           [Parameter(Mandatory = $true)] [string] $ContainerTag,
+           [Parameter(Mandatory = $true)] [string[]] $ArtifactsFolders,
+           [Parameter(Mandatory = $true)] [string] $Registry)
+
+    New-Item -Name $WorkDir\$ContainerSuffix -ItemType directory
+    Compress-Archive -Path $ArtifactsFolders -DestinationPath $WorkDir\$ContainerSuffix\Artifacts.zip
+    # Docker pre 18.03 needs Dockerfile to be within the build context
+    Copy-Item $PSScriptRoot\Dockerfile $WorkDir\$ContainerSuffix
+    # We use ${} to delimit the name before the colon
+    $ContainerName = "contrail-windows-${ContainerSuffix}:$ContainerTag"
+    Invoke-NativeCommand -ScriptBlock {
+        docker build -t $ContainerName $WorkDir\$ContainerSuffix
+    }
+    Invoke-NativeCommand -ScriptBlock {
+        docker tag $ContainerName $Registry/$ContainerName
+    }
+    Invoke-NativeCommand -ScriptBlock {
+        docker push $Registry/$ContainerName
+    }
+}
+
+function Invoke-ContainersBuild {
+    Param ([Parameter(Mandatory = $true)] [string] $WorkDir,
+           [Parameter(Mandatory = $true)] [hashtable[]] $ContainersAttributes,
+           [Parameter(Mandatory = $true)] [string] $ContainerTag,
+           [Parameter(Mandatory = $true)] [string] $Registry)
+
+    $Job.PushStep("Containers build")
+
+    $Job.Step("add insecure registry", {
+        $DockerConfig = @"
+{
+    "insecure-registries" : [ "$Registry" ]
+}
+"@
+        Set-Content -Path "C:\ProgramData\Docker\config\daemon.json" -Value $DockerConfig
```

Comment:
ehh, so what are the ideas we have for this?

Reply:
We can limit a number of executors on each builder to 1.

Reply:
Limiting executors to 1 removes _docker race_ problem in the future. To achieve the same build throughput, we can simply scale the number of builds.

Reply:
Since we (codi) are the ones that setup the insecure registries, can we possibly just use constant number of possible port numbers, like a 100 or something? Then we could generate daemon.json file once with 100 entries.

Reply:
That would lock us to current IP address of the registry machine. Let's go with limiting the number of executors.